### PR TITLE
Github GraphQL Support For On-Prem Github

### PR DESCRIPTION
**Pull request by midnightconman**
This change adds a `github-graphql-endpoint` flag for passing to github graphql client. This mostly fixes 
 github enterprise for tide.

Enhancement on https://github.com/kubernetes/test-infra/pull/11075

components: hook, tide, label_sync, tackle, and prstatus

I am not sure whom to include for approval on this, please make testing and change suggestions.

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
The robots will assign someone useful, that's their job :)

File: prow/prstatus/prstatus.go [hunk 1]
```diff
@@ -37,11 +37,12 @@ import (
...
 
 const (
-	loginSession   = "github_login"
-	githubEndpoint = "https://api.github.com"
-	tokenSession   = "access-token-session"
-	tokenKey       = "access-token"
-	loginKey       = "login"
+	loginSession          = "github_login"
+	githubEndpoint        = "https://api.github.com"
+	githubGraphQLEndpoint = "https://api.github.com/graphql"
```
Comment (stevekuznetsov) [inline prow/prstatus/prstatus.go]: Can we import these from somewhere?

File: prow/github/client.go [hunk 2]
```diff
@@ -226,11 +229,14 @@ func NewClient(getToken func() []byte, bases ...string) *Client {
 //   An endpoint is used when all preceding endpoints have returned a conn err.
 //   This should be used when using the ghproxy GitHub proxy cache to allow
 //   this client to bypass the cache if it is temporarily unavailable.
-func NewDryRunClient(getToken func() []byte, bases ...string) *Client {
+// 'graphqlEndpoint' is the Github GraphQL API (v4) endpoint to use.
+//   The only difference between githubql.NewEnterpriseClient and
```
Comment (stevekuznetsov) [inline prow/github/client.go]: Not sure we need to keep this comment here

Comment (midnightconman) [reply prow/github/client.go]: I'll remove them, no problem.

File: prow/flagutil/github.go [hunk 3]
```diff
@@ -49,7 +51,11 @@ func (o *GitHubOptions) AddFlagsWithoutDefaultGithubTokenPath(fs *flag.FlagSet)
 
 func (o *GitHubOptions) addFlags(wantDefaultGithubTokenPath bool, fs *flag.FlagSet) {
 	o.endpoint = NewStrings("https://api.github.com")
+	o.gitEndpoint = "https://github.com"
```
[the remark this answers is not in the document]
Comment (midnightconman) [reply prow/flagutil/github.go]: I just merged from parent branch (https://github.com/kubernetes/test-infra/pull/11075), the changes should be merged now.

**Comment by stevekuznetsov** [comment]
@midnightconman FYI usually `git rebase` will get you where you want faster than `git merge`

**Comment by midnightconman** [comment]
Going to work on these tests... setting DNM until they are working.

/hold

**Comment by midnightconman** [comment]
I am waiting on https://github.com/kubernetes/test-infra/pull/11214 to keep working on this.

**Comment by midnightconman** [comment]
Hey Guys, starting at the bottom again. This is needed for GitHub Enterprise support. 

As the shurcool library assumes `github.com` if you use `NewClient`, I am using `NewEnterpriseClient`. Please take a look and let me know what you think. It seems the only two apps left not using the centralized `flagutil/github.go` are `label_sync` and `robots/commenter`, so I updated the flags for those.

Thank you again for your time and reviews.

File: prow/github/client.go [hunk 4]
```diff
@@ -231,14 +231,16 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
...
-func NewClientWithFields(fields logrus.Fields, getToken func() []byte, bases ...string) *Client {
+func NewClientWithFields(fields logrus.Fields, getToken func() []byte, graphqlEndpoint string, bases ...string) *Client {
 	return &Client{
 		logger: logrus.WithFields(fields).WithField("client", "github"),
 		time:   &standardTime{},
-		gqlc: githubql.NewClient(&http.Client{
-			Timeout:   maxRequestTime,
-			Transport: &oauth2.Transport{Source: newReloadingTokenSource(getToken)},
-		}),
+		gqlc: githubql.NewEnterpriseClient(
```
Comment (fejta) [inline prow/github/client.go]: What are the side effects of doing this for non-enterprise customers? Should pick which to use more carefully?

Comment (midnightconman) [reply prow/github/client.go]: No side effect, this function wraps `NewClient()` and adds an endpoint url. Here is a link to this function: https://github.com/shurcooL/githubv4/blob/master/githubv4.go#L26

Comment (midnightconman) [reply prow/github/client.go]: I think the function is poorly named, but it is nice they didn't break signature.

File: prow/prstatus/prstatus.go [hunk 5]
```diff
@@ -35,11 +35,12 @@ import (
 )
 
 const (
-	loginSession   = "github_login"
-	githubEndpoint = "https://api.github.com"
-	tokenSession   = "access-token-session"
-	tokenKey       = "access-token"
-	loginKey       = "login"
+	loginSession    = "github_login"
+	githubEndpoint  = "https://api.github.com"
```
Comment (stevekuznetsov) [inline prow/prstatus/prstatus.go]: Can we expose these constants in the GitHub package and use them where appropriate? Would hate to typo `graphql` in a flag default...

**Comment by stevekuznetsov** [comment]
@fejta do we want to get some traction on #12137 first?

File: prow/prstatus/prstatus.go (hunk 5 above)
Comment (midnightconman) [reply prow/prstatus/prstatus.go]: I'm fine with it, hopefully there won't be any cyclical dependency issues. :crossed_fingers:

Comment (midnightconman) [reply prow/prstatus/prstatus.go]: This also kinda touches on the shared flagset conversation we had a week or so ago. Do we want to throw these defaults (constants) in `prow/github` and inherit them in `prow/flagutils/github.go` and in these desperate applications? Or do we want to leave these separate?

**Comment by midnightconman** [comment]
@stevekuznetsov @fejta - PTAL :smile: 

I added centralized defaults for github api endpoint and github graphql endpoint to `prow/github/types.go`... any application using `prow/github` I changed to use these defaults, anything using `vendor/go-github` is still using static strings.

File: prow/flagutil/github.go [hunk 6]
```diff
@@ -66,6 +68,10 @@ func (o *GitHubOptions) Validate(dryRun bool) error {
 		}
 	}
 
+	if _, err := url.Parse(o.graphqlEndpoint); err != nil {
```
Comment (stevekuznetsov) [inline prow/flagutil/github.go]: This one doesn't validate that something _is_ passed

Comment (midnightconman) [reply prow/flagutil/github.go]: Empty would only happen if someone explicitly specified it, the default is covered by the stringvar flag default.

I also think this parse would fail on empty.

Comment (stevekuznetsov) [reply prow/flagutil/github.go]: That's the same as in the other place you have written the check, so let's make them identical code.

Comment (midnightconman) [reply prow/flagutil/github.go]: Ohh, I think I know what you are talking about. I basically copied the existing github endpoint check for each file, so the check logic would be the same. I can update the check logic for github endpoint and graphql endpoint in `flagutil/github.go`, but I would prefer to not update any other files.

This goes back to the flagutil consolidation effort we spoke about a week or two ago. 

Sound good?

Comment (stevekuznetsov) [reply prow/flagutil/github.go]: I would just like to see the content of `Validate()` be the same for every place we're asking for this as a flag

**Comment by midnightconman** [comment]
@cjwagner @stevekuznetsov - PTAL

I updated flagutil to validate we can't have an empty endpoint. Please let me know if you guys need any other changes or have other suggestions.

**Comment by Noahhoetger2001** [comment]
> Going to work on these tests... setting DNM until they are working.
> 
> /hold
> /echo

File: prow/flagutil/github.go (hunk 6 above)
Comment (Noahhoetger2001) [reply prow/flagutil/github.go]: have you tried rerouting it to .shell maybe it might work

Comment (Noahhoetger2001) [reply prow/flagutil/github.go]: another thing. what about the parse code being implemented into something else so it can be passed

Comment (Noahhoetger2001) [reply prow/flagutil/github.go]: /uncc

Comment (Noahhoetger2001) [reply prow/flagutil/github.go]: /uncc
/cc @Noahhoetger2001

**Comment by midnightconman** [comment]
Hi Noah, welcome. Let's let Steve and Cole review this code and we can go from there.

/hold cancel

File: prow/flagutil/github.go [hunk 7]
```diff
@@ -61,11 +63,19 @@ func (o *GitHubOptions) addFlags(wantDefaultGitHubTokenPath bool, fs *flag.FlagS
 // Validate validates GitHub options.
 func (o *GitHubOptions) Validate(dryRun bool) error {
 	for _, uri := range o.endpoint.Strings() {
-		if _, err := url.ParseRequestURI(uri); err != nil {
+		if uri == "" {
+			uri = github.DefaultAPIEndpoint
```
Comment (cjwagner) [inline prow/flagutil/github.go]: The default is already specified above. If `uri == ""` here then `--github-endpoint` was explicitly set to the empty value which should be an error.

File: prow/flagutil/github.go [hunk 8]
```diff
@@ -61,11 +63,19 @@ func (o *GitHubOptions) addFlags(wantDefaultGitHubTokenPath bool, fs *flag.FlagS
...
+		} else if _, err := url.ParseRequestURI(uri); err != nil {
 			return fmt.Errorf("invalid -github-endpoint URI: %q", uri)
 		}
 	}
 
+	if o.graphqlEndpoint == "" {
+		o.graphqlEndpoint = github.DefaultGraphQLEndpoint
```
Comment (cjwagner) [inline prow/flagutil/github.go]: Defaulting is already handled as above.

File: experiment/manual-trigger/manual-trigger.go [hunk 9]
```diff
@@ -55,6 +56,7 @@ func flagOptions() options {
 	flag.StringVar(&o.jenkinsUserName, "jenkins-user-name", "", "Jenkins username.")
 
 	flag.StringVar(&o.githubEndpoint, "github-endpoint", "https://api.github.com", "GitHub's API endpoint.")
+	flag.StringVar(&o.graphqlEndpoint, "graphql-endpoint", "https://api.github.com/graphql", "GitHub's GraphQL API endpoint.")
```
Comment (cjwagner) [inline experiment/manual-trigger/manual-trigger.go]: Pleas use the default value constants.

File: prow/flagutil/github.go (hunk 7 above)
Comment (midnightconman) [reply prow/flagutil/github.go]: I am using the default in validate also, as some testing skips addFlags step, so this value would be empty. Is it possible to have a follow-up PR for this, I have had a bit of back and forth on this requirement from Steve and Fejta. 

The general consensus was to not couple validation logic across individual components.

File: prow/flagutil/github.go (hunk 8 above)
Comment (midnightconman) [reply prow/flagutil/github.go]: Same here, I am explicitly defaulting here, as some individual testing implementations don't run `addFlags`, so this value is empty for those tests.

File: experiment/manual-trigger/manual-trigger.go (hunk 9 above)
Comment (midnightconman) [reply experiment/manual-trigger/manual-trigger.go]: Good catch, switched to use defaults.

**Comment by midnightconman** [comment]
@cjwagner - PTAL

Thank you for your reviews :smile:

**Comment by midnightconman** [comment]
@stevekuznetsov - Cole approved, can I remove hold, or can you, please? :smile:

**Comment by noahwolf2001** [comment]
how long do you think the infra test will be complete
